### ISSUE TITLE
redis execute `rpop` command got block

### ISSUE THREAD

**Issue by zemelLeong** [redis-rs/redis-rs#1041]
Hi. this is my repo: https://github.com/zemelLeong/rbnb20-miner/tree/redis-block-bug

Im trying to get remote redis list data got block when the program run a period time. this is the last part of the log:
```
2024-01-27T23:32:50.202527Z  INFO rbnb20_miner::sender: 开始获取redis数据
2024-01-27T23:32:50.217932Z  INFO rbnb20_miner::sender: 准备提交 "0x64cffffc857cd8914f65eacca3a2959bb1d963e5be0be1a65b99654f78063707"
2024-01-27T23:32:55.574211Z  INFO rbnb20_miner::sender: 状态码: 200 OK
2024-01-27T23:32:55.574755Z  INFO rbnb20_miner::sender: 完成一次数据提交
2024-01-27T23:32:55.575236Z  INFO rbnb20_miner::sender: 开始获取redis数据
2024-01-27T23:32:55.591126Z  INFO rbnb20_miner::sender: 准备提交 "0xdbe05e4897326bf8bc558dfee4dd2c334922f90eb9923916fe2a4eb5ed190731"
2024-01-27T23:32:57.390117Z  INFO rbnb20_miner::sender: 状态码: 200 OK
2024-01-27T23:32:57.390428Z  INFO rbnb20_miner::sender: 完成一次数据提交
2024-01-27T23:32:57.390683Z  INFO rbnb20_miner::sender: 开始获取redis数据
2024-01-27T23:32:57.406502Z  INFO rbnb20_miner::sender: 准备提交 "0x2fcd6865d71f62f6c97eb6d3f403ccbc756cadb612adb53810a3c7a970a25432"
2024-01-27T23:32:57.993004Z  INFO rbnb20_miner::sender: 状态码: 503 Service Unavailable，将重新放入redis或丢弃
2024-01-27T23:32:57.993216Z  INFO rbnb20_miner::sender: 获取redis连接
2024-01-27T23:32:58.049973Z  INFO rbnb20_miner::sender: 执行保存命令
2024-01-27T23:32:58.066267Z  INFO rbnb20_miner::sender: 保存到redis成功
2024-01-27T23:32:58.066548Z  INFO rbnb20_miner::sender: 完成一次数据提交
2024-01-27T23:32:58.066740Z  INFO rbnb20_miner::sender: 开始获取redis数据
2024-01-27T23:32:58.081438Z  INFO rbnb20_miner::sender: 准备提交 "0xb3322f67491d503292940f98f7c9ef6641bb73f1c89d72c203e23f7e7adb90b1"
2024-01-27T23:32:59.206107Z  INFO rbnb20_miner::sender: 状态码: 200 OK
2024-01-27T23:32:59.206434Z  INFO rbnb20_miner::sender: 完成一次数据提交
2024-01-27T23:32:59.206720Z  INFO rbnb20_miner::sender: 开始获取redis数据
2024-01-27T23:32:59.218862Z  INFO rbnb20_miner::sender: 准备提交 "0x51b54f0355a61616b879dba66a6e7e1416fad2fdd6269a7145131cbde8260cfc"
2024-01-27T23:32:59.897034Z  INFO rbnb20_miner::sender: 状态码: 503 Service Unavailable，将重新放入redis或丢弃
2024-01-27T23:32:59.897332Z  INFO rbnb20_miner::sender: 获取redis连接
2024-01-27T23:32:59.949107Z  INFO rbnb20_miner::sender: 执行保存命令
2024-01-27T23:32:59.965745Z  INFO rbnb20_miner::sender: 保存到redis成功
2024-01-27T23:32:59.966020Z  INFO rbnb20_miner::sender: 完成一次数据提交
2024-01-27T23:32:59.966367Z  INFO rbnb20_miner::sender: 开始获取redis数据
2024-01-27T23:32:59.982427Z  INFO rbnb20_miner::sender: 准备提交 "0x7b9590ada1248f0c52d08a9792ffc785029262e8e7a2e4e92455463769cf36bf"
2024-01-27T23:33:01.141166Z  INFO rbnb20_miner::sender: 状态码: 200 OK
2024-01-27T23:33:01.141310Z  INFO rbnb20_miner::sender: 完成一次数据提交
2024-01-27T23:33:01.141456Z  INFO rbnb20_miner::sender: 开始获取redis数据
```
![image](https://github.com/redis-rs/redis-rs/assets/26835087/55ca8771-a136-4a3a-9944-ac8e95c4dca8)

**Comment by nihohit** [comment]
I'm sorry, I don't understand what you're asking / requesting. Can you please expand on what you expected to see and what you've received?

**Comment by zemelLeong** [comment]
The program is fetching data from Redis, and after running for a certain period, it consistently gets stuck at the execution of the `rpop` command. You can refer to the timestamps in the attached log screenshot, where there is a prolonged duration without any output. Prior to this, the log entries were consistently appearing in a cyclic pattern during the program's execution.

**Comment by nihohit** [comment]
try replacing `get_async_connection` with `get_multiplexed_async_connection`. `aio::Connection` is deprecated.

**Comment by zemelLeong** [comment]
Tested once, although it can run for a long time, it seems that there is still a similar problem. I am currently conducting a second test.

**Comment by zemelLeong** [comment]
> try replacing `get_async_connection` with `get_multiplexed_async_connection`. `aio::Connection` is deprecated.

The problem still exists, it took over an hour to run this time.
![image](https://github.com/redis-rs/redis-rs/assets/26835087/80cf506f-1eef-4d47-8cbd-7c301593c3d3)

**Comment by zemelLeong** [comment]
network change `get_multiplexed_async_connection` will get stuck too.
![image](https://github.com/redis-rs/redis-rs/assets/26835087/f03a9df7-b2af-4b7a-b44a-388bc3242ccf)

**Comment by nihohit** [comment]
If you could write me a minimal, locally reproducible test, preferably using only redis-rs' test infrastructure, I could try to debug it. As it is, I don't have enough info to find the source of the issue.

**Comment by zemelLeong** [comment]
I wrote a example. The operating system I'm using for testing is Windows 11, with the Redis service provided by a cloud server. If the CPU usage reaches or exceeds 90 percent, the issue can be reproduced quickly.
![image](https://github.com/redis-rs/redis-rs/assets/26835087/65c79504-39df-4026-a958-9232d5e59a31)

```rust
use redis::AsyncCommands;
use tracing::level_filters::LevelFilter;
use tracing_subscriber::layer::SubscriberExt;
use tracing_subscriber::util::SubscriberInitExt;

struct Sender {
    redis_client: redis::Client,
    queue: String,
}

impl Sender {
    fn init(redis_addr: &str) -> anyhow::Result<Self> {
        let redis_client = redis::Client::open(redis_addr)?;
        Ok(Sender {
            redis_client,
            queue: "test-queue".to_owned(),
        })
    }

    fn get_redis_client(&self) -> redis::Client {
        self.redis_client.clone()
    }

    async fn save_to_redis(&self, data: String) -> anyhow::Result<()> {
        let client = self.get_redis_client();
        tracing::info!("Get redis connection");
        let mut conn = client.get_multiplexed_async_connection().await?;
        tracing::info!("Execute save command");
        conn.lpush(&self.queue, data).await?;
        tracing::info!("Save to redis succeed");

        Ok(())
    }

    async fn run(self) -> anyhow::Result<()> {
        self.save_to_redis("Test Data".to_owned()).await?;
        let client = self.get_redis_client();
        let mut conn = client.get_multiplexed_async_connection().await?;
        loop {
            tracing::info!("Fetching redis data");
            let data: String = conn.rpop(&self.queue, None).await?;
            if data.is_empty() {
                tracing::error!("Data exception");
                break;
            }
            tracing::info!("Data fetch succeed");
            tokio::time::sleep(tokio::time::Duration::from_secs(1)).await;
            self.save_to_redis(data).await?;
            tracing::info!("Finished once data submit");
        }
        Ok(())
    }
}

#[tokio::main]
async fn main() -> anyhow::Result<()> {
    tracing_subscriber::registry()
        .with(LevelFilter::INFO)
        .with(tracing_subscriber::fmt::layer())
        .init();
    let redis_url = "redis://remote-server/";
    let sender = Sender::init(redis_url)?;
    tokio::spawn(sender.run());
    tracing::info!("Thread started");
    loop {
        tokio::time::sleep(tokio::time::Duration::from_secs(10)).await;
    }
}
```

If modify `main` method like this. will be able to see a specific error message `Error: Password authentication failed- AuthenticationFailed`.If the same error occurred during blocking, I don't quite understand why this error message isn't being clearly displayed.
![image](https://github.com/redis-rs/redis-rs/assets/26835087/d292e918-4829-4d64-b3b9-09525b4facb4)

```rust
#[tokio::main]
async fn main() -> anyhow::Result<()> {
    tracing_subscriber::registry()
        .with(LevelFilter::INFO)
        .with(tracing_subscriber::fmt::layer())
        .init();
    let redis_url = "redis://remote-server";
    let sender = Sender::init(redis_url)?;
    sender.run().await?;
    Ok(())
}
```

**Comment by nihohit** [comment]
sorry, this doesn't reproduce the issue when I run it locally against a local redis server.

**Comment by zemelLeong** [comment]
I believe I've found the underlying issue. The high CPU usage, coupled with the unstable network conditions of the remote server, could easily give rise to the `Password authentication failed- AuthenticationFailed` error. It appears that I didn't handle this error properly, which in turn led to the termination of the thread's execution. To address this, I have now implemented additional recovery mechanisms. Thank you.
